interactions/four_particle/__init__.py

```diff
@@ -106,7 +107,8 @@ def integrate(self, ps, stepsize=None):
             self.grids[0].MIN_MOMENTUM / params.aT,
             self.grids[0].MAX_MOMENTUM / params.aT,
             self.grids[1].MIN_MOMENTUM / params.aT,
-            self.grids[1].MAX_MOMENTUM / params.aT
+            self.grids[1].MAX_MOMENTUM / params.aT,
+            self.reaction[3].specie.grid.MAX_MOMENTUM / params.aT
```

Comment:
I'm not sure this should be in here. The problem of extrapolation belongs to the Boltzmann equations setup. Here you internalize it in Boltzmann integrals. So instead of miscounting particles because you did not compute `I_coll` of high enough momenta, now you will underestimate all the collision integrals by limiting the integration region. Am I wrong?

Reply:
This is to ensure that in the reaction 1 + 2+ 3 --> N, the HNL in the final state will have the same momentum as for N --> 1 + 2 + 3 . 
This is necessary, otherwise #particles decayed != #particles created

Reply:
`#particles decayed != #particles created` if you don't include a large enough momentum range. This particular change fixes the superficial discrepancy by making the Boltzmann integral incorrect.

You have to take a large enough momentum range when you setup your system such that `#particles decayed == #particles created`. This is not a problem to fix, this is your flag that all is correct.

Also I bet that the "large enough" momentum is as usual `(m_HNL - m_muon)*a_max`

Reply:
If we have a grid 0 - 20 MeV for HNLs then the HNLs created in 1 + 2 + 3 --> N should at most have momentum of 20 MeV. It just creates a balance.

Reply:
"If we have a grid" is a statement about your discretised code, not about physics. If you want your code to describe physics, you say "if we have HNLs so and so, then we need a grid such and such".

There is something resembling an approximate balance because distribution functions decay exponentially. You are just looking for the sensible cut-off

Reply:
So you suggest increasing MAX_MOMENTUM of HNL grid?

Reply:
You make an estimate of what happens in your system and figure out the numerical parameters. Or make a convergence test increasing the MAX_MOMENTUM until it does not make a difference.

If you change the definition of the Boltzmann integral — you are no longer solving the same equation

Reply:
We integrate only over particles 2 and 3. So in your decay example you don't have to consider the momentum of N at all. I also think that there will be issues with the ordering of particles in reactions if you will start adding exceptions like these.

Simple principle would be better. I propose: "the size of the grid is given by the momenta of particles that can be in the system". If you have a `50 MeV` HNL, it's grid can be tiny, but active neutrinos must have up to `(50 MeV*a)`. If you have a `150 MeV` HNL, you know that there will be `(150-105 MeV)*a` muons. I think this is enough

3T is the mean momentum, not a measure of maximum momentum.